### ISSUE TITLE
goroutine leak

### ISSUE THREAD

**Issue by zhixinwen** [grpc/grpc-go#1936]
Please answer these questions before submitting your issue.

### What version of gRPC are you using?
13975c070286c7371aa3a8b3c230e90d7bf029fc
### What version of Go are you using (`go version`)?
go 1.9.2
### What operating system (Linux, Windows, …) and version?
Linux version 4.4.52 (jenkins@debbuilder02-sjc1) (gcc version 4.9.2 (Debian 4.9.2-10) ) #1 SMP Sun Feb 26 18:02:02 UTC 2017
### What did you do?
We update the our grpc version to `13975c070286c7371aa3a8b3c230e90d7bf029fc`, no other code change. 
p.s we use the experimental `WithBalancer(b Balancer) DialOption` in dial
### What did you expect to see?
nothing should happen
### What did you see instead?
goroutine leaks. Number of goroutine increases with time and does not stop.
We checked the goroutine dump, and see the following goroutine should be the source of leaking.
```
.../vendor/google.golang.org/grpc.newClientStream.func4(0xc420308f00, 0xc424812380, 0x2218ac0, 0xc4223a3ce0)
.../vendor/google.golang.org/grpc/stream.go:298 +0x112
created by code.uber.internal/everything/order/vendor/google.golang.org/grpc.newClientStream
.../vendor/google.golang.org/grpc/stream.go:297 +0xb59
```
We can confirm that we only created one stream ourselves, it is not clear how are other streams created

**Comment by lyuxuan** [comment]
After #1854, we no longer clean up orphan streams. Could you please check whether you explicitly end the stream you created by using the 4 ways described in the #1854? If you didn't end the streams you created accordingly, then the goroutine monitoring the context for each stream will sit there forever, and it might have caused the problem you saw.

If you do end every stream you created accordingly, then there might be a bug on our side. If so, could you please give us a simple reproducible example so we can investigate more? Thank you!

**Comment by menghanl** [comment]
@zhixinwen Do you still see this problem?

**Comment by zhixinwen** [comment]
We rollback the version, and have no further info about the issue.

________________________________

**Comment by menghanl** [comment]
Closing this issue for now. Please reply back or file a new issue if you encounter problems later.
